source/common/http/header_map_impl.h

```diff
@@ -134,20 +127,39 @@ class HeaderMapImpl : public virtual HeaderMap, NonCopyable {
   /**
    * Base class for a static lookup table that converts a string key into an O(1) header.
    */
-  template <class T>
-  struct StaticLookupTable : public TrieLookupTable<StaticLookupResponse (*)(T&)> {
-    using HeaderMapType = T;
-
+  template <class Interface>
+  struct StaticLookupTable
+      : public TrieLookupTable<std::function<StaticLookupResponse(HeaderMapImpl&)>> {
     StaticLookupTable();
 
-    static absl::optional<StaticLookupResponse> lookup(T& header_map, absl::string_view key) {
-      auto entry = ConstSingleton<StaticLookupTable>::get().find(key);
+    void finalizeTable() {
+      CustomInlineHeaderRegistry::finalize<Interface::header_map_type>();
+      auto& headers = CustomInlineHeaderRegistry::headers<Interface::header_map_type>();
+      size_ = headers.size();
+      for (const auto& header : headers) {
+        this->add(header.first.get().c_str(), [&header](HeaderMapImpl& h) -> StaticLookupResponse {
+          return {&h.inlineHeaders()[header.second], &header.first};
+        });
+      }
+    }
+
+    static size_t size() {
+      // The size of the lookup table is finalized when the singleton lookup table is created. This
+      // allows for late binding of custom headers as well as envoy header prefix changes.
```

Comment:
I think this implies that no new headers can be discovered after the singleton is created.

The main problem with this, I guess, is that a route header-matcher looking for a header that's not discovered xDS route configuration cannot be accelerated.

Probably worth calling out the limitation here in this comment if my understanding is correct.

Reply:
> Probably worth calling out the limitation here in this comment if my understanding is correct.

Yes this is correct. I will add more comments. Conceivably it might be possible to have header maps created by an API factory which is owned by the listener/HCM, and then it might be possible to have dynamic acceleration. I will think about this for future work.